### PR TITLE
Fix extra BOS token in front of response for some tokenizers

### PR DESCRIPTION
**Pull request by dakinggg**
Some tokenizers, like llama and mistral, automatically add a BOS token to the start of the input to the tokenizer. This is fine and expected. However, we did not realize that when you call `tokenizer(text='hello', text_target='goodbye')`, it _also_ automatically adds a BOS token to the `labels`. This is unexpected, and would require a user to suffix all of their prompts with BOS after finetuning. So we automatically strip BOS from the labels now for tokenizers that have this behavior.

### CODE REVIEW

File: tests/data/test_template_tokenization.py [hunk 1]
```diff
@@ -163,3 +163,23 @@ def test_tokenize_instruct_example_well_formed():
...
+    # This tokenizer automatically adds bos tokens
+    tokenizer = transformers.AutoTokenizer.from_pretrained(
+        'mistralai/Mixtral-8x7B-v0.1')
+
+    example = {'prompt': 'prompt', 'response': 'response'}
+
+    tokenized_example = tokenize_formatted_example(example, tokenizer)
+
+    assert len(tokenized_example['labels']) == 1
+    assert tokenized_example['labels'][0] != '<s>'
```
Comment (irenedea) [inline tests/data/test_template_tokenization.py:178]: `tokenized_example['labels']` produces a list of int. How about comparing against tokenizer.bos_token_id

```suggestion
    assert len(tokenized_example['labels']) == 1
    assert tokenized_example['labels'][0] != tokenizer.bos_token_id
    assert len(tokenized_example['input_ids'] == 2
    assert tokenized_example['input_ids'][0] == tokenizer.bos_token_id
```

File: llmfoundry/data/finetuning/tasks.py [hunk 2]
```diff
@@ -246,7 +246,15 @@ def _tokenize_prompt_response_formatted_example(
             f'Unable to tokenize example because {response_key} was not a string. {example=}'
         )
 
-    return tokenizer(text=prompt, text_target=response)
+    tokenized_sample = tokenizer(text=prompt, text_target=response)
+
+    # Remove the BOS token from the start of the labels if it was automatically added
+    if hasattr(tokenizer, 'add_bos_token') and tokenizer.add_bos_token:
```
Comment (irenedea) [inline llmfoundry/data/finetuning/tasks.py:252]: we will also need to do this above on L211. pull out into a shared helper fn?

File: tests/data/test_template_tokenization.py [hunk 3]
```diff
@@ -163,3 +163,23 @@ def test_tokenize_instruct_example_well_formed():
             tokenized_example = tokenize_formatted_example(example, tokenizer)
             assert 'input_ids' in tokenized_example
             assert 'labels' in tokenized_example
+
+
+def test_tokenize_no_labels_bos():
+    # This tokenizer automatically adds bos tokens
+    tokenizer = transformers.AutoTokenizer.from_pretrained(
+        'mistralai/Mixtral-8x7B-v0.1')
+
```
Comment (irenedea) [inline tests/data/test_template_tokenization.py:172]:
```suggestion
     assert tokenizer.add_bos_token == True
```

File: tests/data/test_template_tokenization.py [hunk 4]
```diff
@@ -163,3 +163,23 @@ def test_tokenize_instruct_example_well_formed():
...
+    assert len(tokenized_example['labels']) == 1
+    assert tokenized_example['labels'][0] != '<s>'
+
+    # This tokenizer does not have the add_bos_token attribute
+    tokenizer = transformers.AutoTokenizer.from_pretrained('mosaicml/mpt-7b')
+
+    tokenized_example = tokenize_formatted_example(example, tokenizer)
+
+    assert len(tokenized_example['labels']) == 1
```
Comment (irenedea) [inline tests/data/test_template_tokenization.py:185]:
```suggestion
    assert len(tokenized_example['labels']) == 1
    assert tokenized_example['labels'][0] != tokenizer.bos_token_id
    assert len(tokenized_example['input_ids']) == 1
    assert tokenized_example['input_ids'][0] != tokenizer.bos_token_id
```

**Review by irenedea — APPROVED** [review]
Thanks for fixing! Just need to make sure it's done for chat tokenizer as well.

File: tests/data/test_template_tokenization.py (hunk 1 above)
Comment (dakinggg) [reply tests/data/test_template_tokenization.py:178]: oops, thanks

File: llmfoundry/data/finetuning/tasks.py (hunk 2 above)
Comment (dakinggg) [reply llmfoundry/data/finetuning/tasks.py:252]: ah good catch, thanks. will fix

**Review by alextrott16 — COMMENTED** [review]
So, not to be that guy, but I'm not sure that I agree with this behavior. It also conflicts with some of the incoming changes in the [multi-turn chat PR](https://github.com/mosaicml/llm-foundry/pull/985/files/b3e19d1a577aec2ae1202e04e326bfbea7c37552#top).

I am less opinionated about what should happen in the prompt-response formatted case, but when applying the chat template I think the best course of action is to use `tokenize(..., add_special_tokens=False)`. The reason is that the convention is to have the chat template add the special tokens itself into the templated string before tokenizing. Internally in the HF code, when using `apply_chat_template` with `tokenize=True`, you'll see that the method manually turns off adding special tokens when tokenizing because they have already been baked into the templated string. My additions in the linked PR reflect that convention, so I think our bases will be covered in that case.

When it comes to the prompt-response format (where you never use `apply_chat_template`), I think the correct thing to do is less clear. For one, starting the `labels` tokens with BOS is actually correct in the encoder-decoder case. For the decoder-only case, having the first generated token be BOS is really not that bad, IMHO. And it should not change the way that prompts need to be formatted at inference time.

I think the thing that matters most is that our tokenization here reflects the conventions we expect to be applied at inference time.

**Comment by dakinggg** [comment]
Discussed offline, deferring the chat changes to https://github.com/mosaicml/llm-foundry/pull/985

**Review by alextrott16 — APPROVED** [review]
LGTM

**Pull request merged**